File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        long count = 0;
        int left = 0;
        int window = 0;

        int top = Integer.MIN_VALUE;
        for (int n : nums) {
            if (n > top) {
                top = n;
            }
        }

        for (int right = 0; right < nums.length; right++) {
            if (nums[right] == top) {
                window++;
            }

            while (window >= k) {
                if (nums[left] == top) {
                    window--;
                }
                left++;
            }

            count += left;
        }

        return count;
    }
}

File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        # Sliding Window
        # Time Complexity O(n)
        # Space Complexity O(1)

        count = 0
        top = max(nums)
        left = 0
        window = 0

        for right in range(len(nums)):
            if nums[right] == top:
                window += 1
            
            while window >= k:
                if nums[left] == top:
                    window -= 1
                left += 1

            count += left

        return count